oop-object_interactions-medium-q6 (#687)
canvas-gamification/canvas-gamification-junit-tests: src/oop/object_interactions/medium/q6/QuestionText.md

# oop-object_interactions-medium-q6

- Java question

## Question Text

> Companies constantly overbook available seats which causes a lot of passenger frustration. Make a class called 
> **Airplane** and another called **testBooking** with the following descriptions to help with that issue.

Reviewer: ```suggestion
Companies constantly overbook available seats which causes a lot of passenger frustration. 
To solve this problem, airlines sometimes transfer passengers to different flights. 
Make a class called **Airplane** and another called **testBooking** with the following descriptions to help with that issue.
```

## Airplane Class

The `Airplane` class currently has the following items:

- Attributes:
    - An `int` called **maxSeats**
    - An `int` called **numPassengers**
    - None of these attributes should be accessible from other classes.

- Constructors:

>     - A constructor that takes one `int` and initialize the **maxSeats** attribute and initializes the **numPassengers**

Reviewer 2: ```suggestion
    - A constructor that takes one `int` to initialize the **maxSeats** attribute and initializes the **numPassengers**
```

      attribute to zero.
    - A constructor that takes two `int` parameters to initialize the **maxSeats** and **numPassengers** attributes.
    - Both are accessible from other classes.

- Methods:
    - Getter and Setter methods for the **maxSeats** and **numPassengers** attributes.
    - These methods are accessible from other classes.

You must add the following items to the class:

- Methods:

>     - A `transferPassengers` method that take another **Airplane** called **busyPlane** and move some of the

Reviewer 2: ```suggestion
    - A `transferPassengers` method that takes another **Airplane** called **busyPlane** and moves
```

>       passengers of **busyPlane** to this Airplane until this plane is full. If this plane is already full it should

Reviewer: ```suggestion
      passengers of **busyPlane** to this Airplane until it is full. If this plane is already full it should
```

>       write `This plane is full too and cannot accommodate anymore passengers`. If the plane is not full it should write

Reviewer 2: switch write to print

>       `We moved n passengers` if n passengers are moved.

Reviewer: ```suggestion
      `We moved n passengers`  where n is the number of passengers moved.
```

>     - A `toString` method that return the `String` of information in the format of `x out of y seats are filled`. Where

Reviewer 2: ```suggestion
    - A `toString` method that returns a `String` of information in the format of `x out of y seats are filled`. Where
```

    - A constructor that take two `int` parameters to initialize the **maxSeats** and **numPassengers** attributes.
    - Both are accessible from other classes.

- Methods:
    - Getter and Setter methods for **maxSeats** and **numPassengers** attributes
    - These methods are accessible from other classes.

You must add the following items to the class:

- Methods:
    - A `transferPassengers` method that take another **Airplane** called **busyPlane** and move some of the
      passengers of **busyPlane** to this Airplane until this plane is full.
    - If this plane is already full it should write `This plane is full too and cannot accommodate anymore passengers`
    - If the plane is not full it should write `We moved n passengers` if n passengers are moved.
    - A `toString` method that return the `String` of information in the format
      of `myName + " has " + numEyes + " eyes"`
    - All methods must be accessible from other classes.

## TestTransfusion Class

Your `TestBooking` class must have the following items:

- Main Method:
    - A `main` method which initializes three **Airplane** who respectively have 12 seats and 20 passengers, 6 seats and

>       zero passengers, and 4 seats and zero passengers.

Reviewer 2: it's a bit inconsistent to have all the other numbers in numeric form but to spell out 0

    - First move extra passengers from the first plane to second plane. Then move the rest from first to the third
      airplane.

>     - Then print the information of the three **Airplane** in order using **toString**.

Reviewer 2: ```suggestion
    - Then print the information of the three **Airplane** objects in order using **toString**.
```
